pixy/consumer_test.go

```diff
@@ -0,0 +1,543 @@
+
+	sc.Stop()
+}
+
+// If we stop one consumer and start another, the new one picks up where the
+// previous one left off.
+func (s *SmartConsumerSuite) TestSequentialConsume(c *C) {
+	// Given
+	ResetOffsets(c, "group-1", "test.1")
+	produced := GenMessages(c, "sequencial", "test.1", map[string]int{"": 3})
+
+	config := s.newConfig("consumer-1")
+	sc1, err := SpawnSmartConsumer(config)
+	c.Assert(err, IsNil)
+	log.Infof("*** GIVEN 1")
+	consumed := s.consume(c, sc1, "group-1", "test.1", 2)
+	s.assertMsg(c, consumed[""][0], produced[""][0])
+	s.assertMsg(c, consumed[""][1], produced[""][1])
+
+	// When: one consumer stopped and another one takes its place.
+	log.Infof("*** WHEN")
+	sc1.Stop()
+	sc2, err := SpawnSmartConsumer(config)
+	c.Assert(err, IsNil)
+
+	// Then: the second message is consumed.
+	log.Infof("*** THEN")
+	consumed = s.consume(c, sc2, "group-1", "test.1", 1, consumed)
+	s.assertMsg(c, consumed[""][2], produced[""][2])
+	sc2.Stop()
+}
+
+// If a topic has several partitions then they all are being consumed in random
+// order.
+func (s *SmartConsumerSuite) TestMultiPartitionTopic(c *C) {
+	// Given
+	ResetOffsets(c, "group-1", "test.4")
+	GenMessages(c, "multi", "test.4", map[string]int{"A": 100, "B": 100})
+
+	log.Infof("*** GIVEN 1")
+	sc, err := SpawnSmartConsumer(s.newConfig("consumer-1"))
+	c.Assert(err, IsNil)
+
+	// When: exactly one half of all produced events is consumed.
+	log.Infof("*** WHEN")
+	consumed := s.consume(c, sc, "group-1", "test.4", 1)
+	// Wait until first messages from partitions `A` and `B` are fetched.
+	waitFirstFetched(sc, 2)
+	// Consume 100 messages total
+	consumed = s.consume(c, sc, "group-1", "test.4", 99, consumed)
+
+	// Then: we have events consumed from both partitions more or less evenly.
+	log.Infof("*** THEN")
+	if len(consumed["A"]) < 25 || len(consumed["A"]) > 75 {
+		c.Errorf("Consumption disbalance: consumed[A]=%d, consumed[B]=%d", len(consumed["A"]), len(consumed["B"]))
+	}
+
+	sc.Stop()
+}
+
+// When there are more consumers in a group then partitions in a topic then
+// some consumers get assigned no partitions and their consume requests timeout.
+func (s *SmartConsumerSuite) TestTooFewPartitions(c *C) {
+	// Given
+	ResetOffsets(c, "group-1", "test.1")
+	produced := GenMessages(c, "few", "test.1", map[string]int{"": 3})
+
+	sc1, err := SpawnSmartConsumer(s.newConfig("consumer-1"))
+	c.Assert(err, IsNil)
+	log.Infof("*** GIVEN 1")
+	// Consume first message to make `consumer-1` subscribe for `test.1`
+	consumed := s.consume(c, sc1, "group-1", "test.1", 2)
+	s.assertMsg(c, consumed[""][0], produced[""][0])
+
+	// When:
+	log.Infof("*** WHEN")
+	sc2, err := SpawnSmartConsumer(s.newConfig("consumer-2"))
+	c.Assert(err, IsNil)
+	_, err = sc2.Consume("group-1", "test.1")
+
+	// Then: `consumer-2` request times out, when `consumer-1` requests keep
+	// return messages.
+	log.Infof("*** THEN")
+	if _, ok := err.(ErrConsumerRequestTimeout); !ok {
+		c.Errorf("Unexpected error: %s", err)
```

Comment:
I would mention what you are expecting here: `"Expected ErrConsumerRequestTimeout, got %s"`